### PR TITLE
New downloader

### PR DESCRIPTION
**Pull request by KirillTim**
First part of https://github.com/marco-c/crashsimilarity/pull/66

### CODE REVIEW

**Comment by marco-c** [comment]
Can you pull from master now that #73 has been merged?

**Comment by KirillTim** [comment]
Done

File: crashsimilarity/utils.py [hunk 1]
```diff
@@ -12,12 +12,10 @@ def utc_today():
...
 
 
-def get_with_retries(url, params=None, headers=None):
-    retries = Retry(total=16, backoff_factor=1, status_forcelist=[429])
-
+def get_with_retries(url, params=None, headers=None):  # can't be tested with requests_mock.Mocker() for unknown reason
     s = requests.Session()
-    s.mount('https://crash-stats.mozilla.com', HTTPAdapter(max_retries=retries))
-
+    s.mount('https://',
```
Comment (marco-c) [inline crashsimilarity/utils.py]: I think the first parameter of `s.mount` can be `url`.

File: crashsimilarity/utils.py [hunk 2]
```diff
@@ -12,12 +12,10 @@ def utc_today():
...
+            HTTPAdapter(max_retries=Retry(total=5, backoff_factor=1, status_forcelist=[429, 500, 502, 503, 504])))
```
Comment (marco-c) [inline crashsimilarity/utils.py]: I think we should only retry when we get a 429.
Did you get one of 500, 502, 503, 504 with Socorro or Bugzilla?

File: crashsimilarity/downloader.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+from datetime import timedelta
```
Comment (marco-c) [inline crashsimilarity/downloader.py]: I would prefer to keep the Socorro and Bugzilla code separated in two different modules. The Bugzilla module only contains a single function for now, but could become more complex in the future.

File: crashsimilarity/downloader.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        }
+        response = utils.get_with_retries(self._SUPER_SEARCH_URL, params)
+        records = self._json_or_raise(response)['facets']['proto_signature']
+        traces = set([r['term'] for r in records])
+
+        if self._cache:
+            self._cache[key] = traces
+        return traces
+
+    def download_crash_for_id(self, uuid):
```
Comment (marco-c) [inline crashsimilarity/downloader.py]: Nit: rename this `download_crash`.

File: crashsimilarity/download_data.py [hunk 5]
```diff
@@ -10,6 +10,7 @@
 SCHEMA_VERSION = '1'
 
 
+# TODO: rewrite or remove this code
```
Comment (marco-c) [inline crashsimilarity/download_data.py]: Instead of adding *TODO* comments, let's file issues for follow-ups.

File: crashsimilarity/downloader.py (hunk 3 above)
Comment (marco-c) [inline crashsimilarity/downloader.py]: So, could you build a Socorro, Bugzilla and Cache module? The Socorro and Bugzilla modules will use the Cache module.

File: crashsimilarity/utils.py (hunk 2 above)
Comment (KirillTim) [inline crashsimilarity/utils.py]: I didn't. I just try to make function as general as possible. That's why I decide to `s.mount('https://'` only http prefix, but not the concrete url.

File: crashsimilarity/utils.py (hunk 1 above)
Comment (KirillTim) [inline crashsimilarity/utils.py]: Yes, actually.

File: crashsimilarity/utils.py (hunk 2 above)
Comment (marco-c) [inline crashsimilarity/utils.py]: Let's only use 429 for now, as it's the "Too Many Requests" error (so it makes sense to retry after a while). When the server fails, it makes sense for us to fail too.

File: crashsimilarity/downloader.py (hunk 3 above)
Comment (KirillTim) [inline crashsimilarity/downloader.py]: Yes, ok. I will split functions into two classes.

File: crashsimilarity/downloader.py (hunk 4 above)
Comment (KirillTim) [inline crashsimilarity/downloader.py]: Will it be better? 
IMO, it's better to explicitly specify what we are expecting to get, especially in dynamically typed language, isn't it?

File: crashsimilarity/download_data.py (hunk 5 above)
Comment (KirillTim) [inline crashsimilarity/download_data.py]: Ok, I'll remove it and add a new issue.

File: crashsimilarity/downloader.py (hunk 4 above)
Comment (marco-c) [inline crashsimilarity/downloader.py]: In this case, since the only way to get a specific crash from Socorro is via a UUID, the `for_id` is redundant.
It doesn't hurt to have a longer name though, so feel free to leave it as it is (when I say "Nit" it means that it's a minor thing and you can ignore it).

File: crashsimilarity/utils.py (hunk 2 above)
Comment (KirillTim) [inline crashsimilarity/utils.py]: Hm, I agree, will fix.

File: crashsimilarity/downloader.py (hunk 4 above)
Comment (KirillTim) [inline crashsimilarity/downloader.py]: On the other hand, this method will be in Bugzill class. And we exactly know what it suppose to take. Any new contributor will jump to declaration anyway.

**Comment by KirillTim** [comment]
Should I change `downloader_test.py` structure when I split download logic into parts? 
IMO, it's ok to keep one class for all network related testing.

**Comment by marco-c** [comment]
> Should I change downloader_test.py structure when I split download logic into parts?
IMO, it's ok to keep one class for all network related testing.

It's OK for now, as the files are small. We can split it in the future if it grows too large.

**Comment by KirillTim** [comment]
Something goes wrong. To many unrelated changes. Will fix it now

File: crashsimilarity/crash_similarity.py [hunk 6]
```diff
@@ -243,8 +245,15 @@ def signature_similarity(model, paths, signature1, signature2):
 if __name__ == '__main__':
     # download_data.download_crashes(days=7, product='Firefox')
     # paths = download_data.get_paths(days=7, product='Firefox')
-    paths = ['../crashsimilarity_data/firefox-crashes-2016-11-09.json.gz', '../crashsimilarity_data/firefox-crashes-2016-11-08.json.gz', '../crashsimilarity_data/firefox-crashes-2016-11-07.json.gz', '../crashsimilarity_data/firefox-crashes-2016-11-06.json.gz', '../crashsimilarity_data/firefox-crashes-2016-11-05.json.gz', '../crashsimilarity_data/firefox-crashes-2016-11-04.json.gz', '../crashsimilarity_data/firefox-crashes-2016-11-03.json.gz']
+    paths = ['../crashsimilarity_data/firefox-crashes-2016-11-09.json.gz',
```
Comment (KirillTim) [inline crashsimilarity/crash_similarity.py]: I have different Pycharm formatting settings on laptop and desktop. 
Can I keep it as is, without reverting changes back?

**Comment by KirillTim** [comment]
Seems to be done

File: crashsimilarity/crash_similarity.py (hunk 6 above)
Comment (KirillTim) [inline crashsimilarity/crash_similarity.py]: This and other changed code will be removed soon, anyway

**Comment by KirillTim** [comment]
Finally, done :)

File: crashsimilarity/crash_similarity.py (hunk 6 above)
Comment (marco-c) [inline crashsimilarity/crash_similarity.py]: Yes, you can keep this one, but I would prefer the other line-breaking changes in this file to be reverted (I don't like the rules about line length, we should break lines only if it makes them more readable).

File: crashsimilarity/utils.py [hunk 7]
```diff
@@ -34,3 +25,36 @@ def create_dir(path):
...
+    return '{}/{}-crashes-{}.json'.format(CRASHSIMILARITY_DATA_DIR, product.lower(), day)
+
+
+def write_json(path, data):
+    with open(path, 'w') as f:
+        for elem in data:
+            f.write(json.dumps(elem) + '\n')
+
+
+def download_crashes(days, product='Firefox'):
```
Comment (marco-c) [inline crashsimilarity/utils.py]: Can you move this and `get_paths` to SocorroDownloader?

Comment (KirillTim) [inline crashsimilarity/utils.py]: Sure.

**Comment by KirillTim** [comment]
Done

File: crashsimilarity/crash_similarity.py (hunk 6 above)
Comment (marco-c) [inline crashsimilarity/crash_similarity.py]: @KirillTim can you revert the other line-breaking changes in this file, other than this one?

Comment (KirillTim) [inline crashsimilarity/crash_similarity.py]: It is strange. I thought that I'd reverted it. Can't understand what is the problem.
Will make another commit.

Comment (KirillTim) [inline crashsimilarity/crash_similarity.py]: But again the same question here: why should we care about code that will be completely removed?

**Comment by KirillTim** [comment]
Done

File: crashsimilarity/downloader.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                if crash['uuid'] not in uuids:
+                    uuids.add(crash['uuid'])
+                    yield crash
+            if len(crashes) < crashes_per_request:
+                break
+
+    @staticmethod
+    def download_and_save_crashes(days, product='Firefox', save_to_dir=utils.CRASHSIMILARITY_DATA_DIR):
+        if not os.path.exists(save_to_dir):
+            utils.create_dir(save_to_dir)
```
Comment (marco-c) [inline crashsimilarity/downloader.py]: Nit: you can use `utils.create_dir` without checking if the directory exists first (`utils.create_dir` is a no-op if the directory already exists).

File: crashsimilarity/downloader.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        response = self.get_with_retries(self._SUPER_SEARCH_URL, params)
+        records = self._json_or_raise(response)['facets']['proto_signature']
+        traces = set([r['term'] for r in records])
+
+        if self._cache:
+            self._cache[key] = traces
+        return traces
+
+    def download_crash(self, uuid):
+        key = ('crash_for_uuid', uuid, utils.utc_today())
```
Comment (marco-c) [inline crashsimilarity/downloader.py]: Nit: I think the cache here isn't needed (btw, it doesn't change over time, so the key shouldn't contain `utils.utc_today()`)

File: crashsimilarity/utils.py [hunk 10]
```diff
@@ -34,3 +23,16 @@ def create_dir(path):
     except OSError as e:
         if e.errno != errno.EEXIST:
             raise e
+
+
+CRASHSIMILARITY_DATA_DIR = '../crashsimilarity_data'
```
Comment (marco-c) [inline crashsimilarity/utils.py]: Nit: this should go in downloader.py too and probably `crashes_dump_file_path` too.

File: crashsimilarity/downloader.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    _SUPER_SEARCH_URL = 'https://crash-stats.mozilla.com/api/SuperSearch'
+    _PROCESSED_CRASH_URL = 'https://crash-stats.mozilla.com/api/ProcessedCrash'
+
+    def __init__(self, cache=None):
+        super().__init__(cache)
+
+    def download_stack_traces_for_signature(self, signature, traces_num=100, period=timedelta(days=31)):
+        from_date = utils.utc_today() - period
+
+        key = ('traces_for_signature', signature, utils.utc_today(), from_date)
```
Comment (marco-c) [inline crashsimilarity/downloader.py]: Nit: the cache should not depend on the period, so the key should only be `('traces_for_signature', signature, utils.utc_today())`.

File: crashsimilarity/crash_similarity.py (hunk 6 above)
Comment (marco-c) [inline crashsimilarity/crash_similarity.py]: > But again the same question here: why should we care about code that will be completely removed in the very resent(!) future?

It's better to keep a clean history for the future (both for the people who are going to look at the code in the future and for finding regressions). When you look at the history you don't want to see unrelated changes, but want every commit to be clean.
